### PR TITLE
feat(admin): add user create/edit form (UsuarioFormComponent)

### PR DESCRIPTION
**Pull request by FabioCarlesso**
### Motivation
- Replace the placeholder in the admin users area with a working reactive form so administrators can create and edit users from `/admin/usuarios/novo` and `/admin/usuarios/:id/editar` following the project's existing patterns. 
- Keep UI/UX consistent with the rest of the app (validation UX, loading/error states, design tokens and standalone components). 

### Description
- Implemented `UsuarioFormComponent` (`src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts`) as a standalone component using `ReactiveFormsModule`; the component supports create and edit modes, validates `name`, `email`, `password` and `role`, treats password as required on create and optional on edit, and integrates with `UsuarioAdminService` (`listarRoles`, `buscar`, `cadastrar`, `atualizar`).
- Added the form template (`usuario-form.component.html`) with validation messages, loading/error states, a password hint for edit mode and form actions (Save/Cancel), and a small stylesheet (`usuario-form.component.scss`) for the helper text. 
- Extended unit tests (`usuario-form.component.spec.ts`) to cover creation, edition, role-loading fallback, service error handling, invalid route id handling, and navigation after save. 
- Updated documentation to reflect the implemented flow by removing the "placeholder" notes and adding a description of the form in `README.md` and `docs/context.md`.

### Testing
- Ran a development build with `npx ng build --configuration development`, which completed successfully and produced the application bundle. 
- Attempted the full production build with `npm run build`, but it failed due to external font inlining returning HTTP 403 from Google Fonts during the production inline-fonts step. 
- Attempted unit test execution with `npm test -- --watch=false --browsers=ChromeHeadless`; the test bundles were produced but execution failed because the environment lacks a `ChromeHeadless` binary (`CHROME_BIN` not set). 
- Ran static checks (`git diff --check`) with no emitted issues.

------
[Codex Task](https://chatgpt.com/codex/cloud/tasks/task_e_69fd27e08eb48328ba68851ea5302cae)

### CODE REVIEW

File: src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts [hunk 1]
```diff
@@ -1,30 +1,136 @@
...
 
   ngOnInit(): void {
+    this.form = this.fb.group({
+      name: ['', [Validators.required, Validators.minLength(3)]],
+      email: ['', [Validators.required, Validators.email]],
+      password: ['', [Validators.required, Validators.minLength(8)]],
+      role: ['', Validators.required]
+    });
+
+    this.carregarRoles();
```
Comment (FabioCarlesso) [inline src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts]: - **Problem:** `carregarRoles()` is invoked unconditionally on init, even when the route id is invalid (`parametroInvalido = true`) and the form is never rendered.
- **Why it matters:** It triggers an unnecessary HTTP request to `GET /api/users/roles` on a screen the user can't interact with, adding noise to the network tab/server logs and slightly delaying the error message rendering. Also, in edit mode there's a race between `carregarRoles()` and `carregarUsuario()` running in parallel — if `listarRoles()` resolves AFTER `buscar()`, the dropdown options can briefly flicker after the form is populated.
- **Suggested improvement:** Move `carregarRoles()` after the route-param validation, and consider sequencing it before/with `carregarUsuario` (e.g. `forkJoin`) so the role dropdown is guaranteed to contain the loaded user's role before patching the form.
- **Severity:** Medium

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts [hunk 2]
```diff
@@ -1,30 +1,136 @@
...
+  salvando = false;
+  erro: string | null = null;
+  roles: RoleOption[] = ROLE_OPTIONS;
+
+  constructor(
+    private fb: FormBuilder,
+    private service: UsuarioAdminService,
+    private route: ActivatedRoute,
+    private router: Router
+  ) {}
```
Comment (FabioCarlesso) [inline src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts]: - **Problem:** This component switched from the `inject()` pattern (used in the previous version of this file and consistent with other recent components such as `UsuarioListComponent` uses constructor too — but the project elsewhere mixes both) to constructor DI. The PR also removed `private readonly route = inject(ActivatedRoute);` that was already in the file.
- **Why it matters:** Inconsistency hurts readability and makes future refactors (e.g., migrating to standalone signals/inject) more painful. Angular 19 idiomatic style recommends `inject()` for standalone components, especially when the constructor body would otherwise be empty.
- **Suggested improvement:** Prefer `inject()` for all four dependencies to stay consistent with the rest of the file's previous form and Angular 19 conventions:
  ```ts
  private readonly fb = inject(FormBuilder);
  private readonly service = inject(UsuarioAdminService);
  private readonly route = inject(ActivatedRoute);
  private readonly router = inject(Router);
  ```
- **Severity:** Low

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts [hunk 3]
```diff
@@ -1,30 +1,136 @@
...
+
+    if (this.form.invalid || this.salvando) {
+      this.form.markAllAsTouched();
+      return;
+    }
+
+    this.salvando = true;
+    this.erro = null;
+    const valor = this.form.getRawValue();
+    const senha = (valor.password ?? '').trim();
```
Comment (FabioCarlesso) [inline src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts]: - **Problem:** Password is silently trimmed before being sent to the API (`(valor.password ?? '').trim()`), but the input field has no leading/trailing whitespace validator and no UI hint that whitespace will be removed.
- **Why it matters:** A user who legitimately types `"abc1234 "` (8 chars) will pass the `minLength(8)` validator client-side but the server will receive `"abc1234"` (7 chars). This causes silent disagreement between client validation and the value persisted, and breaks future logins if the user thinks the trailing space is part of their password.
- **Suggested improvement:** Either don't trim (let the backend decide), or apply the trim earlier (e.g., on `valueChanges`) so the `minLength(8)` validator runs against the trimmed value the server actually receives. Alternatively add a custom `noWhitespace` validator that rejects passwords with leading/trailing spaces and surfaces a clear error message.
- **Severity:** Medium

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts [hunk 4]
```diff
@@ -1,30 +1,136 @@
...
+          role: valor.role
+        });
+
+    request$.subscribe({
+      next: () => this.router.navigate(['/admin/usuarios']),
+      error: () => {
+        this.erro = 'Erro ao salvar usuário.';
+        this.salvando = false;
+      }
+    });
```
Comment (FabioCarlesso) [inline src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts:98]: - **Problem:** The `error` callback maps every failure to the same generic `"Erro ao salvar usuário."`, regardless of HTTP status code or backend message.
- **Why it matters:** The most common admin-user errors are duplicate email (`409 Conflict`) and validation errors (`400 Bad Request`). The user has no way to distinguish those from a network/500 error and may keep retrying with the same input. It also discards the structured error body that the backend likely returns.
- **Suggested improvement:** Inspect the `HttpErrorResponse` and surface a meaningful message:
  ```ts
  error: (err: HttpErrorResponse) => {
    if (err.status === 409) this.erro = 'Já existe um usuário com este e-mail.';
    else if (err.status === 400) this.erro = err.error?.message ?? 'Dados inválidos.';
    else this.erro = 'Erro ao salvar usuário.';
    this.salvando = false;
  }
  ```
  Apply the same pattern to `carregarUsuario` (e.g., 404 → "Usuário não encontrado.") and add a unit test per status.
- **Severity:** Medium

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts [hunk 5]
```diff
@@ -1,30 +1,136 @@
...
+  private carregarRoles(): void {
+    this.service.listarRoles().subscribe({
+      next: roles => {
+        this.roles = roles.map(role => ({ value: role, label: this.labelRole(role) }));
+      },
+      error: () => {
+        this.roles = ROLE_OPTIONS;
+      }
+    });
+  }
```
Comment (FabioCarlesso) [inline src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts:143]: - **Problem:** When `listarRoles()` succeeds but returns an empty array, `this.roles` becomes `[]` and the `<select>` is left with only the "Selecione..." placeholder. There is no fallback to `ROLE_OPTIONS` for the empty-array case (only for the error case).
- **Why it matters:** A misconfigured backend (or a transient empty response) silently breaks the form — user can't pick a role, can't submit, and there's no error indicator. Edit mode would also drop the user's currently-loaded `role` from the list, so the dropdown won't even show the existing value.
- **Suggested improvement:** Guard the empty case explicitly, e.g.:
  ```ts
  next: roles => {
    this.roles = roles?.length
      ? roles.map(role => ({ value: role, label: this.labelRole(role) }))
      : ROLE_OPTIONS;
  },
  ```
  And cover this with a unit test (mock `listarRoles().and.returnValue(of([]))` and assert `component.roles` falls back).
- **Severity:** Medium

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts [hunk 6]
```diff
@@ -1,30 +1,136 @@
...
     if (this.usuarioId === null) {
       this.parametroInvalido = true;
+      this.erro = 'Identificador inválido.';
       return;
     }
 
     this.isEdit = true;
+    this.form.get('password')?.clearValidators();
+    this.form.get('password')?.setValidators([Validators.minLength(8)]);
+    this.form.get('password')?.updateValueAndValidity();
```
Comment (FabioCarlesso) [inline src/app/pages/admin/usuarios/usuario-form/usuario-form.component.ts:56]: - **Problem:** Three sequential calls (`clearValidators` → `setValidators` → `updateValueAndValidity`) are used to swap `password` validators between create and edit modes. The `clearValidators()` call is redundant because `setValidators()` replaces the validator list.
- **Why it matters:** Minor maintainability — the redundant call obscures the intent and makes a future reader wonder if the order matters. Also, hardcoding the validator swap in `ngOnInit` couples create/edit mode logic.
- **Suggested improvement:** Drop `clearValidators()`:
  ```ts
  this.form.get('password')?.setValidators(Validators.minLength(8));
  this.form.get('password')?.updateValueAndValidity();
  ```
  Or, even cleaner, make `password` validators a function of `isEdit` from the start: `password: ['', this.isEdit ? [Validators.minLength(8)] : [Validators.required, Validators.minLength(8)]]` — but that requires resolving `isEdit` first.
- **Severity:** Low

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/app/pages/admin/usuarios/usuario-form/usuario-form.component.html [hunk 7]
```diff
@@ -3,10 +3,91 @@ <h1>{{ isEdit ? 'Editar Usuário' : 'Novo Usuário' }}</h1>
...
+  </div>
+
+  <div class="form-actions">
+    <button type="submit" class="btn btn-primary" [disabled]="salvando">
+      {{ salvando ? 'Salvando...' : 'Salvar' }}
+    </button>
+    <a routerLink="/admin/usuarios" class="btn btn-outline">Cancelar</a>
   </div>
-}
+</form>
```
Comment (FabioCarlesso) [inline src/app/pages/admin/usuarios/usuario-form/usuario-form.component.html]: - **Problem:** The template was rewritten using the legacy structural directives `*ngIf` / `*ngFor` (and `NgIf`/`NgFor` imports), while the previous version of this same template used Angular's new built-in control flow (`@if` / `@else`). The PR also changed the component imports accordingly.
- **Why it matters:** Angular 19 (the version this project targets per `README.md`) recommends the new control flow — it's faster, requires no module imports, has better type narrowing, and is the project's prior convention in this very file. Mixing the two styles inside the same codebase makes it harder to standardize and migrate.
- **Suggested improvement:** Convert the template to `@if (loading) { ... }`, `@if (!loading && !parametroInvalido) { ... }`, and `@for (role of roles; track role.value) { ... }`, and remove `NgIf`/`NgFor` from the component's `imports`.
- **Severity:** Low

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/app/pages/admin/usuarios/usuario-form/usuario-form.component.html [hunk 8]
```diff
@@ -3,10 +3,91 @@ <h1>{{ isEdit ? 'Editar Usuário' : 'Novo Usuário' }}</h1>
...
+      </select>
+      <div class="invalid-feedback" *ngIf="campo('role')?.invalid && campo('role')?.touched">
+        Perfil é obrigatório.
+      </div>
+    </div>
+
+    <div class="form-group">
+      <label for="password">Senha {{ isEdit ? '' : '*' }}</label>
+      <input
+        id="password"
```
Comment (FabioCarlesso) [inline src/app/pages/admin/usuarios/usuario-form/usuario-form.component.html]: - **Problem:** The required indicator on labels is a literal asterisk character (`Nome *`, `E-mail *`, `Perfil *`, `Senha *`) with no programmatic semantics. In edit mode the asterisk on `Senha` is also removed (`Senha {{ isEdit ? '' : '*' }}`), which means screen readers won't announce the field as required.
- **Why it matters:** Accessibility — assistive tech relies on `aria-required="true"` (or `required` on the input) to announce mandatory fields. Visually-only asterisks are also commonly missed and don't give context to anyone using high-contrast or screen reader modes.
- **Suggested improvement:** Add `aria-required="true"` (or the native `required` attribute, bound to `isEdit` for password) to each required input, and consider wrapping the asterisk in `<span aria-hidden="true">*</span>` so it's purely decorative. Also expose validation messages via `aria-describedby` linking to the `.invalid-feedback` div.
- **Severity:** Low

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/app/pages/admin/usuarios/usuario-form/usuario-form.component.html [hunk 9]
```diff
@@ -3,10 +3,91 @@ <h1>{{ isEdit ? 'Editar Usuário' : 'Novo Usuário' }}</h1>
...
+        [class.is-invalid]="campo('role')?.invalid && campo('role')?.touched"
+      >
+        <option value="">Selecione...</option>
+        <option *ngFor="let role of roles" [value]="role.value">{{ role.label }}</option>
+      </select>
+      <div class="invalid-feedback" *ngIf="campo('role')?.invalid && campo('role')?.touched">
+        Perfil é obrigatório.
+      </div>
+    </div>
+
```
Comment (FabioCarlesso) [inline src/app/pages/admin/usuarios/usuario-form/usuario-form.component.html:82]: - **Problem:** `[autocomplete]="isEdit ? 'new-password' : 'new-password'"` is a no-op ternary — both branches return the same value.
- **Why it matters:** It looks intentional (suggesting that the create vs edit modes were meant to differ), but it doesn't actually do anything, which is confusing for the next reader. It can also be expressed as a static attribute.
- **Suggested improvement:** Replace the binding with the static attribute `autocomplete="new-password"`. If the original intent was to use `current-password` in edit mode, decide explicitly and document it (admins editing other users' passwords typically want `new-password` to avoid password-manager autofill of the admin's own credentials).
- **Severity:** Low

---
_Generated by [Claude Code](https://claude.ai/code)_

File: src/app/pages/admin/usuarios/usuario-form/usuario-form.component.spec.ts [hunk 10]
```diff
@@ -17,52 +41,206 @@ function configure(idParam: string | null) {
...
+      password: 'senha1234',
+      role: 'USER'
+    });
+
+    component.salvar();
+
+    expect(component.erro).toBe('Identificador inválido.');
+    expect(serviceSpy.cadastrar).not.toHaveBeenCalled();
+    expect(serviceSpy.atualizar).not.toHaveBeenCalled();
   });
```
Comment (FabioCarlesso) [inline src/app/pages/admin/usuarios/usuario-form/usuario-form.component.spec.ts:385]: - **Problem:** The spec covers many flows but is missing a few important edge cases:
  - No test that an invalid `email` string (e.g. `"foo"`) sets the `email` control's error and prevents submit.
  - No test that `name` shorter than 3 chars triggers `minlength` and blocks submit.
  - No test that a too-short password in **edit mode** (e.g. `"123"`) fails `minLength(8)` (currently only the create-mode `required` and the empty-edit cases are exercised).
  - No DOM assertions for the rendered validation messages (e.g. "Senha deve ter ao menos 8 caracteres.") after `markAllAsTouched()`.
  - No test for the `salvando` flag preventing a double submit (`salvar(); salvar();` should call `cadastrar` only once).
- **Why it matters:** The validators are the main safeguard before hitting the API; without these tests a future change (e.g. lowering `minLength`) could regress silently. The double-submit test in particular would catch a real bug class for slow networks.
- **Suggested improvement:** Add the cases above; the existing `setup()` helper already supports them.
- **Severity:** Medium

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by FabioCarlesso — COMMENTED** [review]
## Review summary — `feat(admin): add user create/edit form (UsuarioFormComponent)`

Overall this is a well-scoped, well-tested PR that turns the previous placeholder into a working reactive form, with create/edit modes, role loading, error/loading/saving states and a solid spec base. The integration with `UsuarioAdminService` matches the documented backend contracts (`POST /api/users`, `PUT /api/users/{id}`, `GET /api/users/{id}`, `GET /api/users/roles`) and the README/`docs/context.md` were updated consistently. Findings below — inline comments are attached at the relevant lines.

### 1. Documentation
Looks good. `README.md`, `context.md` directory tree, the routes table and the prose section on `/admin` were all updated to remove the "(placeholder)" wording and to describe the new form (fields, roles endpoint, save endpoints, optional password on edit). Item #40 was added to the changelog list and item #37 was rephrased. No stale instructions spotted.

### 2. Bugs and Edge Cases
- See inline: `carregarRoles()` runs even when the route id is invalid and races with `carregarUsuario()` (Medium).
- See inline: empty-array response from `listarRoles()` leaves the dropdown unusable, no `ROLE_OPTIONS` fallback (Medium).
- See inline: silent password trim diverges from the `minLength(8)` validator (Medium).
- Minor: in edit mode `password: ''` is patched into a control with `minLength(8)` validator — empty strings don't trigger `minLength` in Angular, so this is fine, but worth a unit test to lock the behavior.

### 3. Security
Mostly OK. No `[innerHTML]`, no `bypassSecurityTrust*`, no console logs of credentials, password input uses `type="password"`, and `roleGuard(['ADMIN'])` already protects the route upstream. Two notes:
- Password is sent in plaintext over what should be HTTPS (out of scope for this PR, but worth confirming the deployment forces TLS).
- Generic error messages avoid leaking backend internals (good), but see inline about distinguishing 409 (duplicate email) from generic failures.
- **Severity:** Low overall.

### 4. Performance
- No memory-leak risk: `HttpClient` observables auto-complete after one emission, so the missing `takeUntilDestroyed` / unsubscribe is not a real issue here.
- The component does not use `ChangeDetectionStrategy.OnPush`. With a small form and a `campo()` template helper that's invoked many times per CD cycle, this is fine for now but worth considering as an enhancement, especially since other admin components could adopt OnPush together.
- `carregarRoles()` runs even when the form will never be displayed (see inline) — minor wasted request.
- **Severity:** Low.

### 5. Angular Best Practices
- See inline: constructor DI vs `inject()` — the previous version of the file used `inject(ActivatedRoute)`; switching to constructor DI is a step backwards from Angular 19 idiom.
- See inline: legacy `*ngIf`/`*ngFor` instead of `@if`/`@for` (the file's prior version already used the new control flow).
- Validators dynamic swap could be simpler (see inline: `clearValidators` then `setValidators` is redundant).
- Forms could be typed (`fb.nonNullable.group({...})`) for stronger inference on `valor.password` etc.
- `campo(nome: string)` is a thin wrapper that gets called many times from the template — fine, but it returns `AbstractControl | null` with no narrowing; consider exposing typed getters per field or using template variables (`#nameCtrl="ngModel"`-equivalent for reactive forms via `[formControl]`).

### 6. API Integration
- Request/response mapping matches the typed DTOs (`UsuarioAdminCreateRequestDTO`, `UsuarioAdminUpdateRequestDTO`, `UsuarioAdminResponseDTO`) defined in `core/models/usuario-admin.ts`. Good.
- Edit mode correctly omits `password` from the update payload when blank, which aligns with the `UsuarioAdminUpdateRequestDTO`'s optional fields. Good.
- See inline: HTTP status codes are not differentiated in the error path (no 409/400/404 specific messaging).
- The form does not validate uniqueness of `email` client-side (acceptable — that's a server concern), but the UX needs the 409 mapping above to communicate it back.

### 7. UI/UX
- Loading/error/saving states present. Cancel + Save actions present. Field hint on edit-mode password is clear.
- See inline: required-field asterisks are visual-only (accessibility).
- See inline: `[autocomplete]="isEdit ? 'new-password' : 'new-password'"` is a no-op ternary.
- Submit button is only disabled while `salvando`, not when `form.invalid` — that's an intentional, valid UX choice (lets the user click and reveal errors).
- No success toast/snackbar after save — the user just gets navigated to `/admin/usuarios`. Not blocking, but a subtle confirmation message would close the loop.
- No `canDeactivate` guard for unsaved changes — outside scope, but worth tracking as a follow-up across all the project's forms.
- **Severity:** Low.

### 8. Tests
Strong coverage of happy paths (create/edit), service errors (cadastrar/atualizar/buscar), role-loading fallback, invalid id, and submit guard for invalid id. See inline for the gaps (email format, name minlength, password minlength on edit, double-submit, DOM assertions of validation messages).

### 9. Code Quality
- Naming and folder layout match the project (Portuguese method names, English form-control names, files alongside `usuario-list/`).
- `labelRole()` falls back to the raw role string when not found in `ROLE_OPTIONS` — reasonable, but if the backend adds new roles the label won't be localized; consider centralizing the label map alongside `ROLE_OPTIONS`.
- The form template repeats the `campo('x')?.invalid && campo('x')?.touched` pattern; a `showError(name)` helper on the component would dedupe ~6 places.
- Otherwise readable and consistent with the established style.

---

**Net assessment:** Approvable in spirit, but I'd recommend addressing at least the Medium findings (race + ordering of `carregarRoles`, empty-array fallback, password trim, error code differentiation, missing validator tests) before merge. The Low items can be follow-ups.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by FabioCarlesso** [comment]
Apliquei as correções de severidade Medium apontadas no review (commit 7cff3ac):

**`usuario-form.component.ts`**
- `carregarRoles()` deixou de rodar quando o `id` da rota é inválido. Em modo edição, roles e usuário agora carregam juntos via `forkJoin`, eliminando o race condition que podia causar flicker no dropdown e garantindo que o perfil retornado pelo `buscar` exista nas opções antes do `patchValue`.
- Quando `listarRoles()` retorna um array vazio, o componente faz fallback para `ROLE_OPTIONS` (antes só havia fallback no caminho de erro).
- Removido o `.trim()` silencioso da senha — o valor enviado ao backend agora bate exatamente com o validado pelo `minLength(8)` no cliente.
- Erros de salvar/carregar passam a usar `HttpErrorResponse`: 409 → "Já existe um usuário com este e-mail.", 400 → mensagem do backend (ou "Dados inválidos."), 404 no carregar → "Usuário não encontrado.", demais casos mantêm o texto genérico.
- Removido `clearValidators()` redundante antes do `setValidators()` no swap de validadores da senha.

**`usuario-form.component.spec.ts`**
- Novos testes: email inválido bloqueia submit, name abaixo de 3 chars bloqueia submit, senha com menos de 8 chars em edição é inválida, double-submit chama `cadastrar` apenas uma vez, fallback de `ROLE_OPTIONS` quando roles vem `[]`, `listarRoles` não é chamado quando o id da rota é inválido, mapeamento de 409, 400 (com mensagem do backend) e 404 no `buscar`, e preservação de whitespace na senha.

**Validações locais**
- `npx ng build --configuration development` ✅
- `npx tsc --noEmit -p tsconfig.app.json` ✅
- `npx tsc --noEmit -p tsconfig.spec.json` ✅
- `npm test` não pôde ser executado: `CHROME_BIN` não está disponível neste ambiente (mesma limitação reportada no PR original).

Os apontamentos de severidade Low (constructor DI vs `inject()`, migração `*ngIf`/`*ngFor` → `@if`/`@for`, `aria-required`, ternário no-op `[autocomplete]`) ficaram fora deste commit por preferência de escopo — posso aplicá-los em um commit separado se desejar.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by FabioCarlesso** [comment]
Apliquei os ajustes de severidade Low (commit 1d48f5a):

**`usuario-form.component.ts`**
- DI por construtor → `inject()` para `FormBuilder`, `UsuarioAdminService`, `ActivatedRoute` e `Router`, alinhando com Angular 19 e com o padrão prévio do arquivo (`private readonly route = inject(ActivatedRoute);`).
- `imports` do componente: removidos `NgIf` e `NgFor` (não são mais necessários com control flow nativo).

**`usuario-form.component.html`**
- `*ngIf` / `*ngFor` migrados para `@if` / `@for` (com `track role.value` no for de roles).
- Asteriscos de obrigatoriedade movidos para `<span aria-hidden="true">*</span>` — decorativos no DOM, não anunciados por leitores de tela.
- `aria-required="true"` nos campos `name`, `email`, `role`. Em `password`, ligado a `[attr.aria-required]="isEdit ? null : 'true'"` (só obrigatório no cadastro).
- `aria-describedby` em cada input apontando para containers de mensagens (`name-errors`, `email-errors`, `role-errors`, `password-hint password-errors`), com `aria-live="polite"` para que leitores de tela anunciem mudanças nas validações.
- Ternário no-op `[autocomplete]="isEdit ? 'new-password' : 'new-password'"` substituído por atributo estático `autocomplete="new-password"`.
- Banner de erro recebeu `role="alert"`, e o `<form>` ganhou `novalidate` para deixar a validação inteiramente nas mãos do Angular.

**Validações locais**
- `npx ng build --configuration development` ✅
- `npx tsc --noEmit -p tsconfig.app.json` ✅
- `npx tsc --noEmit -p tsconfig.spec.json` ✅
- Os testes da spec continuam passando type-check; `npm test` continua bloqueado por falta de `CHROME_BIN` no ambiente, mesma limitação dos commits anteriores.

---
_Generated by [Claude Code](https://claude.ai/code)_
